sotodlib/io/hkdb.py

```diff
@@ -0,0 +1,425 @@
+    """
+    hk_root: str
+    "Root directory for the HK archive"
+
+    hk_db: str
+    "Path to the hk index database"
+
+    echo_db: bool = False
+    "Whether database operations should be echoed"
+
+    file_idx_lookback_time: Optional[float] = None
+    "Time [sec] to look back when scanning for new files to index"
+
+    show_index_pb: bool = True
+    "If true, shows progress bar when indexing"
+
+    aliases: Dict[str, str] = field(default_factory=dict)
+    """
+    Aliases for hk fields. In this dict, the key is the alias name, and the
+    value is the field descriptor, in the format of ``agent.feed.field``.
+    For example::
+
+        {
+            'fp_temp': 'cryo-ls372-lsa21yc.temperatures.Channel_02_T',
+        }
+    """
+
+    @classmethod
+    def from_yaml(cls, path):
+        with open(path, 'r') as f:
+            return cls(**yaml.safe_load(f))
+
+
+class HkFile(Base):
+    """
+    Database entry for a hk file.
+
+    Args
+    ------
+    path: str
+        Path to the hk file.
+    start_time: float
+        Starting ctime of the file
+    end_time: float
+        Ending ctime of the file
+    size: int
+        Size of file in bytes
+    mod_time: float
+        Modification time of file
+    index_status: float
+        Index status of file. Can be "unindexed", "indexed", or "failed"
+    """
+    __tablename__ = 'hk_files'
+    id = db.Column(db.Integer, primary_key=True)
+    path = db.Column(db.String, nullable=False, unique=True)
+    start_time = db.Column(db.Float)
+    end_time = db.Column(db.Float)
+    size = db.Column(db.Float)
+    mod_time = db.Column(db.Float)
+    index_status = db.Column(db.String)
+
+
+class HkFrame(Base):
+    """
+    Database entry for an hk frame.
+
+    Args
+    --------
+    file_id: int
+        ID of the file containing the frame
+    agent: str
+        instance-id of the OCS agent for the frame
+    feed: str
+        Name of the OCS feed for the frame
+    byte_offset: int
+        Offset of the frame in the G3 file in bytes
+    start_time: float
+        Starting ctime of the frame
+    end_time: float
+        Ending ctime of the frame
+    """
+    __tablename__ = 'hk_frames'
+    id = db.Column(db.Integer, primary_key=True)
+    file_id = db.Column(db.Integer, db.ForeignKey('hk_files.id'))
+    file = relationship('HkFile')
+    agent = db.Column(db.String)
+    feed = db.Column(db.String)
+    byte_offset = db.Column(db.Integer)
+    start_time = db.Column(db.Float)
+    end_time = db.Column(db.Float)
+
+
+class HkDb:
+    """
+    Helper class for createing database sessions
+
+    Args
+    ------
+    cfg : Union[HKConfig, str]
+        Configuration object or path to configuration file
+    """
+    def __init__(self, cfg: Union[HkConfig, str]):
+        if isinstance(cfg, str):
+            cfg = HkConfig.from_yaml(cfg)
+        self.cfg = cfg
+
+        self.engine = db.create_engine(f"sqlite:///{cfg.hk_db}", echo=cfg.echo_db)
```

Comment:
You could generalize this, for non-sqlite options, by accepting engine as a kwarg.

That would facilitate experimentation with alternative databases.

Reply:
This sounds good to me, but instead of passing through an engine I'd rather have the `db_url` settable via the cfg, either as a string or as a [sqlalchemy.url](https://docs.sqlalchemy.org/en/20/core/engines.html#creating-urls-programmatically) object, so it is more straightforward to set via the cfg file, and we don't have to worry about passthrough args. Are there any reasons you think its better for the engine object to be passed in over the URL?

Reply:
Oh, ya, that sounds fine.